nucypher/blockchain/eth/actors.py

```diff
@@ -1643,84 +1653,201 @@ def __init__(self,
         if not self.get_deposited_eth:
             self.BidderError(f"No bids available for {self.checksum_address}")
 
-        # Ensure the claim was not already placed
-        if self._has_claimed:
-            raise self.BidderError(f"Bidder {self.checksum_address} already placed a claim.")
-
         receipt = self.worklock_agent.cancel_bid(checksum_address=self.checksum_address)
         return receipt
 
+    def _get_max_bonus_bid_from_max_stake(self) -> int:
+        """Returns maximum allowed bid calculated from maximum allowed locked tokens"""
+        max_bonus_tokens = self.economics.maximum_allowed_locked - self.economics.minimum_allowed_locked
+        bonus_eth_supply = sum(self._all_bonus_bidders.values()) if self._all_bonus_bidders else self.worklock_agent.get_bonus_eth_supply()
+        bonus_worklock_supply = self.worklock_agent.get_bonus_lot_value()
+        max_bonus_bid = max_bonus_tokens * bonus_eth_supply // bonus_worklock_supply
+        return max_bonus_bid
+
+    def get_whales(self, force_read: bool = False) -> Dict[str, int]:
+        max_bonus_bid_from_max_stake = self._get_max_bonus_bid_from_max_stake()
+
+        bidders = dict()
+        for bidder, bid in self._get_all_bonus_bidders(force_read).items():
+            if bid > max_bonus_bid_from_max_stake:
+                bidders[bidder] = bid
+        return bidders
+
+    def _get_all_bonus_bidders(self, force_read: bool = False) -> dict:
+        if not force_read and self._all_bonus_bidders:
+            return self._all_bonus_bidders
+
+        bidders = self.worklock_agent.get_bidders()
+        min_bid = self.economics.worklock_min_allowed_bid
+
+        self._all_bonus_bidders = dict()
+        for bidder in bidders:
+            bid = self.worklock_agent.get_deposited_eth(bidder)
+            if bid > min_bid:
+                self._all_bonus_bidders[bidder] = bid - min_bid
+        return self._all_bonus_bidders
+
+    def _reduce_bids(self, whales: dict):
+
+        min_whale_bonus_bid = min(whales.values())
+        max_whale_bonus_bid = max(whales.values())
+
+        # first step - align at a minimum bid
+        if min_whale_bonus_bid != max_whale_bonus_bid:
+            whales = dict.fromkeys(whales.keys(), min_whale_bonus_bid)
+            self._all_bonus_bidders.update(whales)
+
+        bonus_eth_supply = sum(self._all_bonus_bidders.values())
+        bonus_worklock_supply = self.worklock_agent.get_bonus_lot_value()
+        max_bonus_tokens = self.economics.maximum_allowed_locked - self.economics.minimum_allowed_locked
+        if (min_whale_bonus_bid * bonus_worklock_supply) // bonus_eth_supply <= max_bonus_tokens:
+            raise self.BidderError(f"Internal error: At least one of bidders {whales} has allowable bid")
+
+        a = min_whale_bonus_bid * bonus_worklock_supply - max_bonus_tokens * bonus_eth_supply
+        b = bonus_worklock_supply - max_bonus_tokens * len(whales)
+        refund = -(-a//b)  # div ceil
```

Comment:
Unsure if there is an alternate (more semantic) approach, but I think this is an acceptable way of performing "ceiling division" (<https://stackoverflow.com/questions/14822184/is-there-a-ceiling-equivalent-of-operator-in-python?noredirect=1&lq=1>)